Introduce flags to specify partition types GUIDs

gjolly opened the pull request:
Instead of only relying on partition type GUIDs to find the ESP and root
filesystem, this allows the user to explicitely specify which partition
UUID they want to use.

jaskaran-khurana commented:
looks good to me

gjolly commented:
I did the following manual testing: https://pastebin.ubuntu.com/p/TTkRM9yryP/ and verified that the image could boot. I used [cvm-tool](https://github.com/gjolly/cvm-tools) to do the annoying vTPM setup and image customization parts.

danpdraper commented on deploy.go:
```diff
@@ -49,9 +49,10 @@ type deployOptions struct {
-	AzDiskProfile string `long:"az-disk-profile" description:""`
-	UefiConfig    string `long:"uefi-config" description:"JSON file describring the platform firmware configuration"`
+	RootPartitionUUID        string `long:"root-partition-uuid" description:"Explicitly provide the root partition UUID instead of relying on partition type"`
+	ESPPartitionUUID         string `long:"efi-partition-uuid"  description:"Explicitly override the ESP partition UUID instead of relying on partition type"`
```
Redundant space before `description`

`Explicitly provide` instead of `Explicitly override` (for consistency with the line above)

danpdraper commented on encrypt.go:
```diff
@@ -57,6 +57,9 @@ type encryptOptions struct {
 	GrowRoot            bool   `long:"grow-root" description:"Grow the root partition to fill the available space, disabling cloud-init's cc_growpart"`
 
+	RootPartitionUUID string `long:"root-partition-uuid" description:"Explicitly provide the root partition UUID instead of relying on partition type"`
+	ESPPartitionUUID  string `long:"efi-partition-uuid"  description:"Explicitly override the ESP partition UUID instead of relying on partition type"`
```
Redundant space before `description`

`Explicitly provide` instead of `Explicitly override` (for consistency with the line above)

danpdraper commented on deploy.go:
```diff
@@ -467,6 +468,17 @@ func (d *imageDeployer) run(opts *deployOptions) error {
+	espPartitionUUIDOverride := opts.ESPPartitionUUID != ""
+
+	if rootPartitionUUIDOverride != espPartitionUUIDOverride {
+		return errors.New("Please override both root and efi partititions")
```
The language should suggest that both overridden and neither overridden are the two acceptable states; the current wording only captures the former.

danpdraper commented on encrypt.go:
```diff
@@ -505,6 +507,16 @@ func (e *imageEncrypter) run(opts *encryptOptions) error {
+	espPartitionUUIDOverride := opts.ESPPartitionUUID != ""
+
+	if rootPartitionUUIDOverride != espPartitionUUIDOverride {
+		return errors.New("Please override both root and efi partititions")
```
The language should suggest that both overridden and neither overridden are the two acceptable states; the current wording only captures the former.

danpdraper commented on internal/gpt/gpt_test.go:
```diff
@@ -0,0 +1,53 @@
+		Parts         Partitions
+		ExpectedEntry *PartitionEntry
+	}{
+		{"symple", partition3.UniquePartitionGUID.String(), Partitions{partition1, partition2, partition3}, partition3},
```
`simple` instead of `symple`, although I will suggest `match` instead of `simple` to reduce ambiguity

danpdraper commented on internal/gpt/gpt_test.go:
```diff
@@ -0,0 +1,53 @@
+	}{
+		{"symple", partition3.UniquePartitionGUID.String(), Partitions{partition1, partition2, partition3}, partition3},
+		{"wrong UUID", "foobar", Partitions{partition1}, nil},
+		{"no parition", partition1.UniquePartitionGUID.String(), Partitions{}, nil},
```
`partition` instead of `parition`

danpdraper commented on internal/gpt/gpt_test.go:
```diff
@@ -0,0 +1,53 @@
+		{"symple", partition3.UniquePartitionGUID.String(), Partitions{partition1, partition2, partition3}, partition3},
+		{"wrong UUID", "foobar", Partitions{partition1}, nil},
+		{"no parition", partition1.UniquePartitionGUID.String(), Partitions{}, nil},
+		{"no parition matching", partition1.UniquePartitionGUID.String(), Partitions{partition2, partition3}, nil},
```
`partition` instead of `parition`

danpdraper commented on internal/gpt/gpt_test.go:
```diff
@@ -0,0 +1,53 @@
+		t.Run(testParams.Name, func(t *testing.T) {
+			partition := testParams.Parts.FindByUUID(testParams.UUIDStr)
+			if partition != testParams.ExpectedEntry {
+				t.Error("Wrong partition entry returned by FindByUUID")
```
It would be helpful if the error message included `testParams.Parts` and `testParams.UUIDStr`.

gjolly replied:
This is how it looks now:

```
gpt_test.go:49: Wrong partition entry returned by FindByUUID(0067b62e-04b6-5a80-04d6-66e51bac31fa) on [EFI_PARTITION_ENTRY{ PartitionTypeGUID: 00000000-0000-0000-0000-000000000000, UniquePartitionGUID: af6bb61f-621a-0115-9786-a8307070adeb, StartingLBA: 0x0, EndingLBA: 0x0, Attributes: 0x0000000000000000, PartitionName: "" } EFI_PARTITION_ENTRY{ PartitionTypeGUID: 00000000-0000-0000-0000-000000000000, UniquePartitionGUID: af0b5bbd-87e8-3d22-4d41-4d95059eda3e, StartingLBA: 0x0, EndingLBA: 0x0, Attributes: 0x0000000000000000, PartitionName: "" } EFI_PARTITION_ENTRY{ PartitionTypeGUID: 00000000-0000-0000-0000-000000000000, UniquePartitionGUID: 0067b62e-04b6-5a80-04d6-66e51bac31fa, StartingLBA: 0x0, EndingLBA: 0x0, Attributes: 0x0000000000000000, PartitionName: "" }]. Got EFI_PARTITION_ENTRY{ PartitionTypeGUID: 00000000-0000-0000-0000-000000000000, UniquePartitionGUID: 0067b62e-04b6-5a80-04d6-66e51bac31fa, StartingLBA: 0x0, EndingLBA: 0x0, Attributes: 0x0000000000000000, PartitionName: "" }, expected EFI_PARTITION_ENTRY{ PartitionTypeGUID: 00000000-0000-0000-0000-000000000000, UniquePartitionGUID: af0b5bbd-87e8-3d22-4d41-4d95059eda3e, StartingLBA: 0x0, EndingLBA: 0x0, Attributes: 0x0000000000000000, PartitionName: "" }.
```

I can't make it way more readable because I don't know if `partition` or `testParams.ExpectedEntry` are `nil` or not and I don't really want to add extra logic here to test that.

danpdraper commented on main.go:
```diff
@@ -215,27 +215,44 @@ func (b *encryptCloudImageBase) connectNbd(path string) error {
+	if rootPartitionUUID != "" {
+		root = partitions.FindByUUID(rootPartitionUUID)
+		if root == nil {
+			return fmt.Errorf("cannot find partition by partition uuid %v on %s ", rootPartitionUUID, b.devPath)
```
Superfluous space following `%s`

danpdraper commented on main.go:
```diff
@@ -215,27 +215,44 @@ func (b *encryptCloudImageBase) connectNbd(path string) error {
+			return fmt.Errorf("cannot find partition by partition uuid %v on %s ", rootPartitionUUID, b.devPath)
+		}
+	} else {
+		// XXX: Could there be more than one partition with this type?
```
Please replace `XXX` with either `TODO` or an issue number.

gjolly committed:
Add tests for Partitions.FindByUUID

gjolly reviewed (commented):
If think I fixed all your comments @danpdraper.

gjolly committed:
Introduce flags to specify partition UUIDs

Instead of only relying on partition type GUIDs to find the ESP and root
filesystem, this allows the user to explicitely specify which partition
UUID they want to use.
